crates/sui-indexer/src/handlers/checkpoint_handler_v2.rs

```diff
@@ -0,0 +1,817 @@
+    tx_indexing_receiver: mysten_metrics::metered_channel::Receiver<TemporaryCheckpointStoreV2>,
+    commit_notifier: watch::Sender<Option<CheckpointSequenceNumber>>,
+) where
+    S: IndexerStoreV2 + Clone + Sync + Send + 'static,
+{
+    use futures::StreamExt;
+
+    info!("Indexer checkpoint commit task started...");
+    let checkpoint_commit_batch_size = std::env::var("CHECKPOINT_COMMIT_BATCH_SIZE")
+        .unwrap_or(5.to_string())
+        .parse::<usize>()
+        .unwrap();
+    info!("Using checkpoint commit batch size {checkpoint_commit_batch_size}");
+
+    let mut stream = mysten_metrics::metered_channel::ReceiverStream::new(tx_indexing_receiver)
+        .ready_chunks(checkpoint_commit_batch_size);
+
+    while let Some(indexed_checkpoint_batch) = stream.next().await {
+        let mut checkpoint_batch = vec![];
+        let mut tx_batch = vec![];
+        let mut events_batch = vec![];
+        let mut tx_indices_batch = vec![];
+        let mut object_changes_batch = vec![];
+        let mut packages_batch = vec![];
+        let mut _epoch_batch = None;
+
+        if config.skip_db_commit {
+            info!(
+                "[Checkpoint/Tx] Downloaded and indexed checkpoint {:?} - {:?} successfully, skipping DB commit...",
+                indexed_checkpoint_batch.first().map(|c| c.checkpoint.sequence_number),
+                indexed_checkpoint_batch.last().map(|c| c.checkpoint.sequence_number),
+            );
+            continue;
+        }
+
+        for indexed_checkpoint in indexed_checkpoint_batch {
+            let TemporaryCheckpointStoreV2 {
+                checkpoint,
+                transactions,
+                events,
+                tx_indices,
+                object_changes,
+                packages,
+                epoch,
+            } = indexed_checkpoint;
+            checkpoint_batch.push(checkpoint);
+            tx_batch.push(transactions);
+            events_batch.push(events);
+            tx_indices_batch.push(tx_indices);
+            object_changes_batch.push(object_changes);
+            packages_batch.push(packages);
+            _epoch_batch = epoch;
+        }
+
+        let first_checkpoint_seq = checkpoint_batch.first().as_ref().unwrap().sequence_number;
+        let last_checkpoint_seq = checkpoint_batch.last().as_ref().unwrap().sequence_number;
+
+        let guard = metrics.checkpoint_db_commit_latency.start_timer();
+        let tx_batch = tx_batch.into_iter().flatten().collect::<Vec<_>>();
+        let tx_indices_batch = tx_indices_batch.into_iter().flatten().collect::<Vec<_>>();
+        let events_batch = events_batch.into_iter().flatten().collect::<Vec<_>>();
+        let packages_batch = packages_batch.into_iter().flatten().collect::<Vec<_>>();
+        let checkpoint_num = checkpoint_batch.len();
+        let tx_count = tx_batch.len();
+
+        // TODO: persist epoch
+
+        futures::future::join_all(vec![
+            state.persist_transactions(tx_batch),
```

Comment:
with GIN, each tx also has one row in the tx_indices table, what's your thought of having separate tables for tx and its indices <> one tx table with indices?